### DISCUSSION TITLE
novnc - HTML5 vnc viewer

### DISCUSSION THREAD

**Discussion by spikefishjohn** [GNS3/gns3-web-ui#1358] [Ideas]
What was the downside to this again? I didn't find much discussion on it accept for this.

I founds this commit but didn't really understand what happened. Looks like it was opened as an enhancement and then just closed?

**Comment by grossmj** [comment]
Where did you find it?

**Comment by spikefishjohn** [comment]
oh sorry, I didn't link what I found. It wasn't a commit, it looks like it was a enhancement request?

https://github.com/GNS3/gns3-web-ui/issues/412

**Comment by ariweisz** [comment]
Can I add my voice to this as a feature request? I've been discussing on the community and novnc works fine with GNS3 images launched through qemu with everything specified as an advanced config. It could be a really neat way to provide a completely self-contained server version of GNS3 when used with the web client. 

https://www.gns3.com/influencing-qemu-vnc-syntax-for-use-with-novnc

**Reply by spikefishjohn** [reply]
Post the advanced options you are using and explain any gotchas you’ve found and how to set it up. 

Nice going! Looking forward to seeing how you did that.

**Reply by ariweisz** [reply]
What I did was simple, which is why I hope it will be an easy feature. 

Using a Linux VM inside qemu (let's pretend the IP is 10.10.10.10), I copied the launch syntax: 

/usr/local/bin/qemu-system-x86_64 -name fw-mgmt -m 2048M -smp cpus=1,sockets=1 -enable-kvm -machine smm=off -boot order=c -drive file=/home/gns3/GNS3/projects/3c13b5f1-ca84-416c-b84c-2c4c19f98f9b/project-files/qemu/5dae2878-6258-4c41-868c-f32a1d317ec6/hda_disk.qcow2,if=virtio,index=0,media=disk,id=drive0 -uuid 5dae2878-6258-4c41-868c-f32a1d317ec6 -vnc 10.10.10.10:0 -monitor tcp:127.0.0.1:17227,server,nowait -net none -device virtio-net-pci,mac=0c:8f:9b:7e:c6:00,netdev=gns3-0 -netdev socket,id=gns3-0,udp=127.0.0.1:20015,localaddr=127.0.0.1:20014 &

then I tried it by inserting a websocket listener on 5910 with ",websocket=5910": 

/usr/local/bin/qemu-system-x86_64 -name fw-mgmt -m 2048M -smp cpus=1,sockets=1 -enable-kvm -machine smm=off -boot order=c -drive file=/home/gns3/GNS3/projects/3c13b5f1-ca84-416c-b84c-2c4c19f98f9b/project-files/qemu/5dae2878-6258-4c41-868c-f32a1d317ec6/hda_disk.qcow2,if=virtio,index=0,media=disk,id=drive0 -uuid 5dae2878-6258-4c41-868c-f32a1d317ec6 -vnc 10.10.10.10:0**,websocket=5910** -monitor tcp:127.0.0.1:17227,server,nowait -net none -device virtio-net-pci,mac=0c:8f:9b:7e:c6:00,netdev=gns3-0 -netdev socket,id=gns3-0,udp=127.0.0.1:20015,localaddr=127.0.0.1:20014 &

when that worked, I went back to GNS3 and turned off the console, then set the advanced options: 
 -vga virtio -vnc 10.10.10.10:0,websocket=5910

so I think it means it's enough just to add the ",websocket=xxxx" string to get qemu to attach a websocket for VNC. 

On the novnc side, I just installed httpd then pulled the novnc source into the default location: 

cd /var/www/html/
git clone https://github.com/novnc/noVNC
chmod -R 755 /var/www/html/novnc

then I connect to http://10.10.10.10/novnc/vnc.html 
in the novnc options I removed the string "websockify" (line 220 in vnc.html)

I think the more complicated part is having the web UI generate a console link on the websocket port instead of the console port.

**Comment by josephmhiggins** [comment]
@ariweisz,
Oh, what you are saying is that if an advanced user explicitly types in a configuration for an option (normal gns3 users do not do this), then your explicit configuration parameters override the parameters received from the normal user's tabs. Yeah, that looks cool to me, although I have no say in the matter. I like it.

**Reply by spikefishjohn** [reply]
This really isn’t a very helpful. The reason I’m asking for the specifics is so @grossmj has the details needed to maybe add this feature. 

GNS3 is in bad need of a pure web based console and this would be a great step in the right direction.
